Improve stable library apis per Scott's feedback

janeyx99 opened the pull request:
Following 3 suggestions:
1. inline at::Tensor arg
2. use uniq ptr of array vs std::vector
3. document the `std::optional<S>()` case

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #152040

janeyx99 committed:
Improve stable library apis per Scott's feedback

[ghstack-poisoned]

swolchok commented on docs/source/notes/libtorch_stable_abi.md:
```diff
@@ -9,6 +9,7 @@ This note will eventually contain more details on how to use the APIs in torch/c
 |  type in custom extension    |   StableIValue representation   |   type in libtorch  |   Schema Type  |
 | -------- | ------- | ------- | ------- |
 | std::optional\<S> | raw bitwise copy into leading bytes of uint64_t of pointer to a new StableIValue representing S | std::optional\<T> | Type? |
+| std::optional\<S>() | nullptr | std::optional\<T>() | Type? |
 | std::nullopt | nullptr | IValue() | None |
```
The third column here doesn't track for me -- isn't it IValue(std::optional\<T>()), [which is equivalent to IValue()](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/core/ivalue_inl.h#L2308)? I would expect lines 12 and 13 here to be redundant (and I guess I would advise removing line 13).

swolchok commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1433,23 +1432,23 @@ class StableIValueBoxedKernel : public c10::OperatorKernel {
     const auto num_arguments = schema.arguments().size();
 
-    std::vector<StableIValue> ministack(std::max(num_arguments, num_returns));
+    std::unique_ptr<StableIValue[]> ministack(new StableIValue[std::max(num_arguments, num_returns)]);
```
nit: `auto ministack = std::make_unique<StableIValue[]>(std::max(num_arguments, num_returns));`

swolchok commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1433,23 +1432,23 @@ class StableIValueBoxedKernel : public c10::OperatorKernel {
       const auto ministack_idx = num_arguments - idx - 1;
       const c10::TypePtr& arg_type = schema.arguments()[ministack_idx].type();
-      ministack[ministack_idx] = from_ivalue(arg_type, torch::jit::pop(stack));
+      ministack.get()[ministack_idx] = from_ivalue(arg_type, torch::jit::pop(stack));
```
nit: you don't need to change this line because of https://en.cppreference.com/w/cpp/memory/unique_ptr/operator_at

swolchok commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1433,23 +1432,23 @@ class StableIValueBoxedKernel : public c10::OperatorKernel {
     for (size_t idx = 0; idx < num_returns; idx++) {
       const c10::TypePtr& ret_type = schema.returns()[idx].type();
-      torch::jit::push(stack, to_ivalue(ret_type, ministack[idx]));
+      torch::jit::push(stack, to_ivalue(ret_type, ministack.get()[idx]));
```
ditto don't need to change

swolchok reviewed (approved):
accepted with minor complaints

janeyx99 committed:
Update on "Improve stable library apis per Scott's feedback"

Following 3 suggestions:
1. inline at::Tensor arg
2. use uniq ptr of array vs std::vector
3. document the `std::optional<S>()` case




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

janeyx99 committed:
Update on "Improve stable library apis per Scott's feedback"

Following 3 suggestions:
1. inline at::Tensor arg
2. use uniq ptr of array vs std::vector
3. document the `std::optional<S>()` case




[ghstack-poisoned]

albanD reviewed (approved):
SGTM !
